overlord/devicestate/devicemgr.go

```diff
@@ -528,6 +538,143 @@ func (m *DeviceManager) ensureBootOk() error {
 	return nil
 }
 
+func (m *DeviceManager) ensureCloudInitRestricted() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	if m.cloudInitAlreadyRestricted {
+		return nil
+	}
+
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// On Ubuntu Core devices that have been seeded, we want to restrict
+	// cloud-init so that its more dangerous (for an IoT device at least)
+	// features are not exploitable after a device has been seeded. This allows
+	// device administrators and other tools (such as multipass) to still
+	// configure an Ubuntu Core device on first boot, and also allows cloud
+	// vendors to run cloud-init with only a specific data-source on subsequent
+	// boots but disallows arbitrary cloud-init {user,meta,vendor}-data to be
+	// attached to a device via a USB drive and inject code onto the device.
+
+	// TODO: expand the scope to run on any device with a gadget, so i.e.
```

Comment:
we should delete this TODO, I will do this in a followup with some other cleanups that we didn't do at the time